app-libs/stf/src/stf_sgx.rs

```diff
@@ -152,23 +166,171 @@ where
 
 	fn execute_call(
 		state: &mut State,
+		shard: &ShardIdentifier,
 		call: TCS,
 		calls: &mut Vec<ParentchainCall>,
 		node_metadata_repo: Arc<NodeMetadataRepository>,
 	) -> Result<(), Self::Error> {
-		state.execute_with(|| call.execute(calls, node_metadata_repo))
+		state.execute_with(|| call.execute(calls, shard, node_metadata_repo))
 	}
 
-	fn on_initialize(state: &mut State, now: Moment) -> Result<(), Self::Error> {
+	fn on_initialize(
+		state: &mut State,
+		shard: &ShardIdentifier,
+		integritee_block_number: BlockNumber,
+		now: Moment,
+	) -> Result<(), Self::Error> {
 		trace!("on_initialize called at epoch {}", now);
 		state.execute_with(|| {
 			// as pallet_timestamp doesn't export set_timestamp in no_std, we need to re-build the same behaviour
 			sp_io::storage::set(&storage_value_key("Timestamp", "Now"), &now.encode());
 			sp_io::storage::set(&storage_value_key("Timestamp", "DidUpdate"), &true.encode());
 			<Runtime::OnTimestampSet as OnTimestampSet<_>>::on_timestamp_set(now.into());
+			// ensure we're assuming the correct storage encoding based on pallet version
+			if get_mirrored_parentchain_storage_by_key_hash::<u16>(
+				<EnclaveBridgeStorage as EnclaveBridgeStorageKeys>::pallet_version(),
+				&ParentchainId::Integritee,
+			)
+			.map_or(false, |v| v <= 1)
+			{
+				if let Some(config) = get_mirrored_parentchain_storage_by_key_hash(
+					EnclaveBridgeStorage::upgradable_shard_config(shard),
+					&ParentchainId::Integritee,
+				) {
```

Comment:
the way it works now is:
1. mirror some keys in opaque manner
2. interpret values in on_initialize and perform necessary actions.

I'm not entirely sure we will keep it this way, becasue we're effectively double-storing values here. Yet, it seemed to me that DevX is better if we can handle L1 state mirroring purely from app-libs, which means we must handle it opaquely outside of STF